core/src/bytecode/typecheck/record.rs

```diff
@@ -0,0 +1,543 @@
+//! Typechecking records.
+//!
+//! Because record literal definitions are flexible in Nickel (piecewise definitions), they need
+//! a bit of preprocessing before they can be typechecked. Preprocessing and typechecking of
+//! records is handled in this module.
+use super::*;
+use crate::{
+    bytecode::ast::record::{FieldDef, FieldPathElem, Record},
+    combine::Combine,
+    position::TermPos,
+};
+
+use std::iter;
+
+use indexmap::{map::Entry, IndexMap};
+
+pub(super) trait Resolve<'ast> {
+    type Resolved;
+
+    fn resolve(&'ast self) -> Self::Resolved;
+}
+
+/// A resolved record literal, without field paths or piecewise definitions. Piecewise definitions
+/// of fields have been grouped together, path have been broken into proper levels and top-level
+/// fields are partitioned between static and dynamic.
+#[derive(Default)]
+pub(super) struct ResolvedRecord<'ast> {
+    /// The static fields of the record.
+    pub stat_fields: IndexMap<LocIdent, ResolvedField<'ast>>,
```

Comment:
Maybe it's not important in this context, but in principle there are multiple locations for each field, right? It looks like `ResolvedRecord::combine` is just choosing one. (The LSP certainly wants them all, but I guess it isn't using this resolution anyway...)

Reply:
Indeed. For now I'm just maintaining the old behavior, which was to use `TermPos::None` as soon as the parser had to merge multiple definitions either statically or dynamically, precisely because the `RichTerm` interface requires that you provide at most one position, which isn't possible. If you look at `ResolvedField::pos`,it uses an XOR so that we return a defined position iff there is only one position defined; if there is none or there are multiple ones, we bail out and return `None`.

We can certainly do better in the future, I suppose. Although having a multi-pieces definitions in a statically typed block is currently un-ergonomic enough (since we can't type merging correctly, everything has to be of type `Dyn`) that I don't expect it to happen much. Still, I expect to re-use the record resolution part for compilation, where you might want all the positions. And one day we might type `&` more precisely as well.